src/core/common/time_ticker.hpp

```diff
@@ -0,0 +1,136 @@
+
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "common/time.hpp"
+#include "common/timer.hpp"
+
+namespace ot {
+
+/**
+ * This class represents a time ticker.
+ *
+ * The time ticker emits periodic ticks (with 1 second period interval) to a set of registered tick receiver modules.
+ * The tick receivers (OpenThread objects) are identified by the `Receiver` enumeration. The receiver objects
+ * must provide `HandleTimeTick()` method which would be invoked by `TimeTicker` periodically.
+ *
+ */
+class TimeTicker : public InstanceLocator, private NonCopyable
+{
+public:
+    /**
+     * This enumeration represents time tick receivers.
+     *
+     * This enumeration contains the list of all OpenThread modules that can be registered as time tick receivers.
+     *
+     */
+    enum Receiver : uint8_t
+    {
+#if OPENTHREAD_FTD
+        kMleRouter,       ///< `Mle::MleRouter`
+        kAddressResolver, ///< `AddressResolver`
+#if OPENTHREAD_CONFIG_CHILD_SUPERVISION_ENABLE
+        kChildSupervisor, ///< `Utils::ChildSupervisor`
+#endif
+#endif
+#if OPENTHREAD_CONFIG_IP6_FRAGMENTATION_ENABLE
+        kIp6FragmentReassembler, ///< `Ip6::Ip6` (handling of fragmented messages)
+#endif
+#if OPENTHREAD_CONFIG_DUA_ENABLE || OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        kDuaManager, ///< `DuaManager`.
+#endif
+#if OPENTHREAD_CONFIG_MLR_ENABLE || OPENTHREAD_CONFIG_TMF_PROXY_MLR_ENABLE
+        kMlrManager, ///< `MlrManager`
+#endif
+        kMeshForwarder, ///< `MeshForwarder`
+    };
+
+    /**
+     * This constructor initializes the `TimeTicker` instance.
+     *
+     */
+    TimeTicker(Instance &aInstance);
+
+    /**
+     * This method registers a receiver with `TimeTicker` to receive periodic ticks.
+     *
+     * @param[in] aReceiver   A tick receiver identifier.
+     *
+     */
+    void RegisterReceiver(Receiver aReceiver);
+
+    /**
+     * This method unregisters a receiver with `TimeTicker` to receive periodic ticks.
+     *
+     * @param[in] aReceiver   A tick receiver identifier.
+     *
+     */
+    void UnregisterReceiver(Receiver aReceiver);
+
+    /**
+     * This method indicates whether a receiver is registered with `TimeTicker` to receive periodic ticks.
+     *
+     * @param[in] aReceiver   A tick receiver identifier.
+     *
+     * @retval TRUE   If @p aReceiver is registered with `TimeTicker`.
+     * @retval FALSE  If @p aReceiver is not registered with `TimeTicker`.
+     *
+     */
+    bool IsReceiverRegistered(Receiver aReceiver) const { return (mReceivers & Mask(aReceiver)) != 0; }
+
+private:
+    static_assert(kMeshForwarder <= 32, "Too many `Receiver`s - does not fit in a `uint32_t` bit mask");
+
+    enum
+    {
+        kTickInterval = 1000, // in msec.
+    };
+
+    constexpr static uint32_t Mask(Receiver aReceiver) { return static_cast<uint32_t>(1U) << aReceiver; }
+
+    static void HandleTimer(Timer &aTimer);
+    void        HandleTimer(void);
+
+    uint32_t   mReceivers;
```

Comment:
nit: use `uint16_t` since it's enough for now?

Reply:
Even if we define it `uint16_t`, due to added padding (for alignment of `Timer`) 4 bytes will be used. So better to use the 4 bytes  in our definition.

Also changing the order having `Timer` first and then `mReceiver` second won't help, since `Timer` itself is `InstanceLocator` and the main class `TimeTicker` is also `InstanceLocator`, and even though the `InstanceLocator` is empty base class, some toolchains would add padding at start to ensure the two `InstanceLocator` get different addresses).

Reply:
Fine to me